fetch_index.py

```diff
@@ -369,53 +557,148 @@ def indexing(self, collection):
             "Successfuly fetched all the objects for site collection: %s"
             % (collection)
         )
+
         logger.info(
             "Saving the checkpoint for the site collection: %s" % (collection)
         )
-        if not self.is_error:
-            self.checkpoint.set_checkpoint(collection, current_time)
 
+        is_error_shared.append(self.is_error)
 
-def start():
+        for obj in ['sites', 'lists', 'list_items']:
+            if ids.get(obj):
+                prev_ids = storage[obj]
+                prev_ids.update(ids[obj])
+                storage[obj] = prev_ids
+
+
+def datetime_partitioning(start_time, end_time, processes):
+    """ Divides the timerange in equal partitions by number of processors
+        :param start_time: start time of the interval
+        :param end_time: end time of the interval
+        :param processes: number of processors the device have
+    """
+    start_time = datetime.strptime(start_time, DATETIME_FORMAT)
+    end_time = datetime.strptime(end_time, DATETIME_FORMAT)
+
+    diff = (end_time - start_time) / processes
+    for idx in range(processes):
+        yield (start_time + diff * idx)
+    yield end_time
+
+
+def start(indexing_type):
+    """Runs the indexing logic regularly after a given interval
+        or puts the connector to sleep
+        :param indexing_type: The type of the indexing i.e. Incremental Sync or Full sync
+    """
     logger.info("Starting the indexing..")
     config = Configuration("sharepoint_connector_config.yml", logger)
 
+    is_error_shared = multiprocessing.Manager().list()
+
     if not config.validate():
         print_and_log(
             logger,
             "error",
             "Terminating the indexing as the configuration parameters are not valid",
         )
         exit(0)
-
-    indexing_interval = 60
+    data = config.reload_configs()
     while True:
-        data = config.reload_configs()
+        current_time = (datetime.utcnow()).strftime("%Y-%m-%dT%H:%M:%SZ")
+        ids_collection = {"global_keys": {}}
+        storage_with_collection = {"global_keys": {}, "delete_keys": {}}
+
+        if (os.path.exists(IDS_PATH) and os.path.getsize(IDS_PATH) > 0):
+            with open(IDS_PATH) as ids_store:
+                try:
+                    ids_collection = json.load(ids_store)
+                except ValueError as exception:
+                    logger.exception(
+                        "Error while parsing the json file of the ids store from path: %s. Error: %s"
+                        % (IDS_PATH, exception)
+                    )
+            # delete all the permissions present in workplace search
+            if indexing_type == "full_sync":
```

Comment:
Why this is needed for full sync?

Reply:
Hi, this class acts as a wrapper, it calls the start method of fetch_index and passes a parameter indexing_type to denote the full sync. The user can run this file at a fixed interval, that can be specified in the yml file

Reply:
Sorry, I didn't make it clear. What I meant is, why is `removing all permissions` needed for full sync?

Reply:
@wangch079 , I can take this one. The reason we are removing all the permissions is that we do not have an update method in permissions, hence the only way to sync permissions is to delete all the existing permissions and bring in the latest ones

Reply:
Actually you can either add permissions atop the existing ([doc](https://www.elastic.co/guide/en/workplace-search/master/workplace-search-document-permissions-api.html#add-one), [code](https://github.com/elastic/enterprise-search-python/blob/1788413218badc01e2da23ac290698de40117f8c/elastic_enterprise_search/client/_workplace_search.py#L1416)), or completely overwrite ([doc](https://www.elastic.co/guide/en/workplace-search/master/workplace-search-document-permissions-api.html#add-all), [code](https://github.com/elastic/enterprise-search-python/blob/1788413218badc01e2da23ac290698de40117f8c/elastic_enterprise_search/client/_workplace_search.py#L1531))

Reply:
@wangch079  The reason for not using this put_permissions was to handle the below scenario:

1. Consider there is a user1 in Sharepoint and the permissions for that user had been synced in the workplace
2. Now the user1 is deleted from the Sharepoint
3. As the put_permissions indexes the permissions per user, and in the user's list from Sharepoint we will no longer get user1 in the response, the permission for user1 will not be reflected in the Workplace

Hence, the put_permissions would work fine when the permissions are updated for a user but if the user is entirely deleted, the put_permissions will not sync the permissions for that user and they would still be able to search the documents in the Workplace. Let me know if we are missing things here.

Reply:
Hi @nirav-elastic , let's leave this aside as this is not related to the PR.

Reply:
Okay, @wangch079 . Let us know if there is anything else to be covered in this PR before you can approve it